File: motoman/src/com/marcowong/motoman/ObjLoaderSkeletonPatcher.java
package com.marcowong.motoman;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Mesh;
import com.badlogic.gdx.graphics.Pixmap;
import com.badlogic.gdx.graphics.VertexAttribute;
import com.badlogic.gdx.graphics.VertexAttributes;

public class ObjLoaderSkeletonPatcher {
	public final static Color[] skeletonIdColors = new Color[] {
		Color.WHITE, Color.BLACK, Color.RED, Color.ORANGE, Color.YELLOW, Color.GREEN, Color.CYAN,
		Color.BLUE, Color.MAGENTA, Color.PINK, Color.LIGHT_GRAY, Color.GRAY, Color.DARK_GRAY
	};
	
	public void patch(ModelData model, Pixmap skeletonMapping) {
		for (SubMeshData subMesh : model.subMeshes) {
			MeshData mesh = subMesh.mesh;
			int sizeVertex = 3 + (mesh.hasNorms ? 3 : 0) + (mesh.hasUVs ? 2 : 0);
			int nVertex = mesh.vertices.length / sizeVertex;
			float[] vertices = mesh.vertices;
			float[] vertices2 = new float[nVertex * (sizeVertex + 1)];
			
			int uvOffset = 3 + (mesh.hasNorms ? 3 : 0);
			
			for (int i = 0; i < nVertex; ++i) {
				int offset = i * sizeVertex;
				int offset2 = i * (sizeVertex + 1);
				int offsetUV = offset2 + uvOffset;
				int offsetSke = offset2 + sizeVertex;
				for (int j = 0; j < sizeVertex; ++j) vertices2[offset2 + j] = vertices[offset + j];
				float u = 0;
				float v = 0;
				if (mesh.hasUVs) {
					u = vertices2[offsetUV];
					v = vertices2[offsetUV + 1];
				}
				vertices2[offsetSke] = getSkeletonId(skeletonMapping, u, v);
			}
			
			mesh.vertices = vertices2;
			mesh.hasSkeleton = true;
		}
	}
	
	private Color tmpColor = new Color();
	private int getSkeletonId(Pixmap m, float u, float v) {
		int x = (int)Math.round(u * m.getWidth());
		int y = (int)Math.round(v * m.getHeight());
		Color.rgba8888ToColor(tmpColor, m.getPixel(x, y));
		int id = 0;
		float lowestDiff = Float.POSITIVE_INFINITY;
		for (int i = 0; i < skeletonIdColors.length; ++i) {
			float diff = 
				Math.abs(skeletonIdColors[i].r - tmpColor.r) +
				Math.abs(skeletonIdColors[i].g - tmpColor.g) +
				Math.abs(skeletonIdColors[i].b - tmpColor.b);
			if (diff < lowestDiff) {
				id = i;
				lowestDiff = diff;
			}
		}
		return id;
	}
}

File: motoman/src/commonMain/kotlin/com/marcowong/motoman/model/ObjLoaderSkeletonPatcher.kt
package com.marcowong.motoman.model

import com.marcowong.motoman.gl.Pixmap
import com.marcowong.motoman.track.math.Color
import kotlin.math.abs
import kotlin.math.roundToInt

class ObjLoaderSkeletonPatcher {
    companion object {
        val skeletonIdColors = arrayOf(
            Color(1f, 1f, 1f, 1f), // WHITE
            Color(0f, 0f, 0f, 1f), // BLACK
            Color(1f, 0f, 0f, 1f), // RED
            Color(1f, 0.647f, 0f, 1f), // ORANGE
            Color(1f, 1f, 0f, 1f), // YELLOW
            Color(0f, 1f, 0f, 1f), // GREEN
            Color(0f, 1f, 1f, 1f), // CYAN
            Color(0f, 0f, 1f, 1f), // BLUE
            Color(1f, 0f, 1f, 1f), // MAGENTA
            Color(1f, 0.686f, 0.686f, 1f), // PINK
            Color(0.75f, 0.75f, 0.75f, 1f), // LIGHT_GRAY
            Color(0.5f, 0.5f, 0.5f, 1f), // GRAY
            Color(0.25f, 0.25f, 0.25f, 1f) // DARK_GRAY
        )
    }

    fun patch(model: ModelData, skeletonMapping: Pixmap) {
        for (subMesh in model.subMeshes) {
            val mesh = subMesh.mesh ?: continue
            val sizeVertex = 3 + (if (mesh.hasNorms) 3 else 0) + (if (mesh.hasUVs) 2 else 0)
            val nVertex = if (sizeVertex > 0) mesh.vertices.size / sizeVertex else 0
            val vertices = mesh.vertices
            val vertices2 = FloatArray(nVertex * (sizeVertex + 1))

            val uvOffset = 3 + (if (mesh.hasNorms) 3 else 0)

            for (i in 0 until nVertex) {
                val offset = i * sizeVertex
                val offset2 = i * (sizeVertex + 1)
                val offsetUV = offset2 + uvOffset
                val offsetSke = offset2 + sizeVertex
                for (j in 0 until sizeVertex) {
                    vertices2[offset2 + j] = vertices[offset + j]
                }
                var u = 0f
                var v = 0f
                if (mesh.hasUVs) {
                    u = vertices2[offsetUV]
                    v = vertices2[offsetUV + 1]
                }
                vertices2[offsetSke] = getSkeletonId(skeletonMapping, u, v).toFloat()
            }

            mesh.vertices = vertices2
            mesh.hasSkeleton = true
        }
    }

    private fun getSkeletonId(m: Pixmap, u: Float, v: Float): Int {
        val x = (u * m.width).roundToInt()
        val y = (v * m.height).roundToInt()
        // clamp x and y to pixmap bounds
        val cx = x.coerceIn(0, m.width - 1)
        val cy = y.coerceIn(0, m.height - 1)
        val pixel = m.getPixel(cx, cy)
        val r = ((pixel ushr 24) and 0xFF) / 255f
        val g = ((pixel ushr 16) and 0xFF) / 255f
        val b = ((pixel ushr 8) and 0xFF) / 255f

        var id = 0
        var lowestDiff = Float.POSITIVE_INFINITY
        for (i in skeletonIdColors.indices) {
            val color = skeletonIdColors[i]
            val diff = abs(color.r - r) + abs(color.g - g) + abs(color.b - b)
            if (diff < lowestDiff) {
                id = i
                lowestDiff = diff
            }
        }
        return id
    }
}
